File: CSharp/HtmlWalker/ITag.cs
using HtmlWalker.Core;
using HtmlWalker.Platform;
using HtmlWalker.Tag.Base;
using HtmlWalker.Visitor;

namespace HtmlWalker
{
    /// <summary>
    ///   Describes what the inner tag types are
    /// </summary>
    /// <author>Robert R Flores</author>
    public enum ContentType {
        /// <summary>
        ///   No content will be returned by <see cref="ITag.OwnedTags"/> or <see cref="ITag.TextValue"/>.
        ///   Identifies <see cref="ContentNoneTag"/>
        /// </summary>
        NONE,

        /// <summary>
        ///   Only string text available by <see cref="ITag.TextValue"/>.
        ///   Identifies <see cref="ContentTextTag"/> and <see cref="ContentNoneTag"/>
        /// </summary>
        TEXT,

        /// <summary>
        ///   Multiple tags available by <see cref="ITag.OwnedTags"/>.
        ///   No content from <see cref="ITag.TextValue"/>
        ///   Identifies <see cref="ContentContainerTag"/>
        /// </summary>
        TAG,

        /// <summary>
        ///   Only script text available by <see cref="ITag.TextValue"/>.
        ///   Text may be formatted differently than <see cref="ContentType.TEXT"/>.
        ///   Identifies <see cref="ContentScriptTag"/>
        /// </summary>
        SCRIPT,

        /// <summary>
        ///   For what is not defined as a normal HTML tag.
        ///   Whether there is <see cref="ITag.TextValue"/> or <see cref="ITag.OwnedTags"/>
        ///   depends on each subclass.
        ///   Identifies <see cref="ContentUtilTag"/>
        /// </summary>
        UTIL,

        /// <summary>
        ///   This tag is a place holder for when a tag is marked to be skipped
        ///   from processing.
        ///   Identifies <see cref="ContentUtilTag"/>
        /// </summary>
        SKIP,
    }

    /// <summary>
    ///   Interface describing an HTML tab
    /// </summary>
    /// <author>Robert R Flores</author>
    public interface ITag
    {
        /// <summary>
        ///  Call for the Vistor pattern
        /// </summary>
        /// <param name="visitor">vister object</param>
        void Visit(ITagVoidVisitor visitor);

        /// <summary>
        ///   Call for a Visitor pattern using Generics
        /// </summary>
        /// <typeparam name="T">generic type</typeparam>
        /// <param name="visitor">vister object</param>
        /// <returns></returns>
        T Visit<T>(ITagGenericVisitor<T> visitor);

        /// <summary>
        ///   Access to any <see cref="ITag"/> owned by this tag
        /// </summary>
        ITag Owner { get; }

        /// <summary>
        /// Returns the enum <b>ContentType</b> for this tag adapter
        /// </summary>
        ContentType ContentType { get; }

        /// <summary>
        ///  Returns the HTML tag object housed by this tag adapter
        /// </summary>
        object Tag { get; }

        /// <summary>
        ///  Returns the tag name as in 'html', 'body', 'p' ...
        /// </summary>
        string Name { get; }

        /// <summary>
        ///   Gets/Sets the string content for tags that do not contain other tags
        ///   Applies to tags of content type<b>TEXT</b> and <b>SCRIPT</b>
        /// </summary>
        string TextValue { get; set; }

        /// <summary>
        ///   Access to the tags owned by this tag
        /// </summary>
        TagList OwnedTags { get; }

        /// <summary>
        ///   The sum of all the <see cref="TextValue"/> of the <see cref="OwnedTags"/>
        /// </summary>
        string OwnedText { get; }

        /// <summary>
        ///   Access to the HTML tag attributes
        /// </summary>
        TagAttributes Attributes { get; }

        /// <summary>
        ///   Creates a shallow clone of the adapter with a cloned HTML tag
        /// </summary>
        /// <param name="factory">factory used to create the clone</param>
        /// <param name="owner">tag to be atteched to. Can be null</param>
        /// <returns></returns>
        ITag Clone(WalkerPlatform.WalkerFactory factory, ITag owner);
    }
}

File: Java/src/htmlwalker/ITag.java
/*
 * 
 */
package htmlwalker;

import htmlwalker.core.TagAttributes;
import htmlwalker.exception.HtmlWalkerException;
import htmlwalker.platform.WalkerFactory;
import htmlwalker.tag.base.TagList;
import htmlwalker.visitor.ITagGenericVisitor;
import htmlwalker.visitor.ITagVoidVisitor;

/**
 * This interface provides a non-platform specific representation of an HTML tag.
 * 
 * The intention is that this interface will be impleneted by tag adapter classes
 * for each type of HTML tag. 
 * 
 * @author Robert Flores
 */
public interface ITag
{
	
	/**
	 * This method provides the implementation of the visitor pattern for the ITagVoidVisitor.
	 *
	 * @param visitor - the visitor
	 * @throws HtmlWalkerException the html walker exception
	 * 
	 * @see HtmlWalker.ITagVoidVisitor
	 */
	public void visit(ITagVoidVisitor visitor) throws HtmlWalkerException;
	
	/**
	 * This method provides the implementation of the visitor pattern for the ITagGenericVisitor.
	 *
	 * @param <T> - the generic type that is returned by each 'visit'
	 * @param visitor - the visitor
	 * @return T
	 * @throws HtmlWalkerException the html walker exception
	 * 
	 * @see HtmlWalker.ITagGenericVisitor
	 */
	public <T> T visit(ITagGenericVisitor<T> visitor) throws HtmlWalkerException;
	
	/**
	 * This enum represents what kinds of tags will be return as content.
	 */
	public enum ContentType
	{
		NONE, /** No content will be returned by <b>innerTag</b> and <b>innerText</b> */
		TEXT, /** Only string text available by <b>innerText</b> */
		TAG, /** Multiple tags available by <b>innerTag</b> */
		SCRIPT, /** Only script text available by <b>innerText</b> */
		UTIL, /** For what is normally not an HTML tag. The subclass defines <b>innerTag</b> and <b>innerText</b> */
		SKIP, /** Acting as a place holder for tags is to be skipped from processing. */
	}
	
	/**
	 * Returns the enum <b>ContentType</b> for this tag adapter
	 *
	 * @return the content type 
	 * 
	 * @see HtmlWalker.ITag.ContentType
	 */
	public ContentType contentType();
	
	/**
	 * Access to the HTML tag object housed by this tag adapter.
	 *
	 * @return the tag object
	 */
	public Object tag();
	
	/**
	 * Returns the tag name.
	 *
	 * @return the string
	 */
	public String name();
	
	public ITag owner();
	
	public TagAttributes attributes();
	
	/**
	 * Gets the string content for tags that do not contain other tags.
	 * 
	 * Applies to tags of content type <b>TEXT</b> and <b>SCRIPT</b>
	 *
	 * @return the content text
	 */
	public String getTextValue();
	
	/**
	 * Sets the string content for tags that do not contain other tags.
	 * 
	 * Applies to tags of content type <b>TEXT</b> or <b>SCRIPT</b>
	 *
	 * @param text the new content text
	 */
	public void setTextValue(String text);
	
	public TagList ownedTags();
	
	public String ownedText();
	
	public ITag clone(WalkerFactory factory, ITag owner);
}
